File: packages/promotion-gate-python/promotion_gate/chain.py
"""Environment promotion chain logic.

Defines the ordered promotion path that a service version must follow:
local -> dev -> qa -> uat -> staging -> production -> live

UAT variants (uat1, uat2, uat3) are normalised to "uat" when checking
the chain position.
"""

from __future__ import annotations

from typing import List, Optional

CHAIN: List[str] = ["local", "dev", "qa", "uat", "staging", "production", "live"]

UAT_VARIANTS: List[str] = ["uat", "uat1", "uat2", "uat3"]


def normalize(env: str) -> str:
    """Normalise an environment name to its canonical chain form.

    UAT variants (uat1, uat2, uat3) are collapsed to ``"uat"``.
    All names are lower-cased and stripped of whitespace.
    """
    env = env.strip().lower()
    if env in UAT_VARIANTS:
        return "uat"
    return env


def _index_of(env: str) -> int:
    """Return the index of *env* in the chain, or raise ValueError."""
    norm = normalize(env)
    try:
        return CHAIN.index(norm)
    except ValueError:
        raise ValueError(
            f"Unknown environment: {env!r}. Valid environments: {CHAIN}"
        )


def previous_of(env: str) -> Optional[str]:
    """Return the environment that must contain an active promotion before
    a service can be promoted into *env*.

    Returns ``None`` if *env* is the first in the chain.
    """
    idx = _index_of(env)
    if idx <= 0:
        return None
    return CHAIN[idx - 1]


def is_unrestricted(env: str) -> bool:
    """Return ``True`` when the environment does not require a prior
    promotion record (i.e. ``"local"``).
    """
    return normalize(env) == "local"


def uat_variants() -> List[str]:
    """Return a copy of the accepted UAT sub-environment names."""
    return list(UAT_VARIANTS)

File: packages/promotion-gate-node/src/chain.ts
/**
 * Environment promotion chain and helpers.
 *
 * The canonical order a version must traverse before reaching production:
 *   dev -> qa -> uat -> staging -> production
 *
 * UAT is a single logical stage with multiple parallel variants
 * (uat1, uat2, uat3). Any one of them satisfies the "uat" gate.
 */

/** Ordered list of logical environment stages. */
export const CHAIN: readonly string[] = [
  'dev',
  'qa',
  'uat',
  'staging',
  'production',
] as const;

/** Concrete UAT variant names that map to the logical "uat" stage. */
export const UAT_VARIANTS: readonly string[] = [
  'uat1',
  'uat2',
  'uat3',
] as const;

/**
 * Normalize an environment name to its logical chain stage.
 * e.g. "uat2" -> "uat", "production" -> "production"
 */
export function normalize(env: string): string {
  const lower = env.toLowerCase().trim();
  if (UAT_VARIANTS.includes(lower)) {
    return 'uat';
  }
  return lower;
}

/**
 * Return the stage that must have been promoted *before* `env` is reachable.
 * Returns `undefined` for "dev" (nothing precedes it).
 */
export function previousOf(env: string): string | undefined {
  const stage = normalize(env);
  const idx = CHAIN.indexOf(stage);
  if (idx <= 0) {
    return undefined;
  }
  return CHAIN[idx - 1];
}

/**
 * Returns `true` when no promotion gate applies (i.e. the first stage).
 */
export function isUnrestricted(env: string): boolean {
  return normalize(env) === CHAIN[0];
}

/**
 * Expand a logical stage to its concrete variants.
 * "uat" -> ["uat1","uat2","uat3"], anything else -> [env]
 */
export function uatVariants(env: string): string[] {
  if (normalize(env) === 'uat') {
    return [...UAT_VARIANTS];
  }
  return [env.toLowerCase().trim()];
}
